### PR TITLE
Fix global typings for extensions

### LINKED ISSUE #2963: Globals variables ($, dayjs) are not correctly type-hinted
**Workaround:**

Copy [`@types` folder](https://github.com/flarum/core/tree/master/js/%40types) from core.

### PR DESCRIPTION
**Pull request by davwheat**
**Fixes #2963**

**Changes proposed in this pull request:**

- Adds support for our global typings in extensions
- Linked tsconfig PR sets up this configuration
- Drops global `app` typings (people should import instead)
- Adds common namespace `app` export
- Adds missing `app` imports to core
- Converts Mithril imports within core to type imports (causes error if we attempt to use the imported mithril for anything except types as this causes duplicated code)
- Adds typings for global `flarum` variable

**Reviewers should focus on:**

To test this locally...
1. Require this branch in an extension's `composer.json`
2. `composer update`
3. `cd vendor/flarum/core/js`
4. `npm i`
5. `npm run build-typings`
6. Modify extension's `tsconfig.json` as per https://github.com/flarum/flarum-tsconfig/pull/2
7. Test global typings! (dayjs, mithril)

**Screenshot**

**Confirmed**

- [x] Frontend changes: tested on a local Flarum installation.
- [ ] Backend changes: tests are green (run `composer test`).

**Required changes:**

- [ ] Related documentation PR: https://github.com/flarum/docs/pull/358
- [ ] Related core extension PRs: https://github.com/flarum/flarum-tsconfig/pull/2

### CODE REVIEW

File: js/src/@types/global.d.ts [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+// Extend JQuery with our custom functions, defined with $.fn
+interface JQuery {
+  /**
+   * Flarum's tooltip JQuery plugin.
+   *
+   * Do not use this directly. Instead use the `<Tooltip>` component that
+   * is exported from `flarum/common/components/Tooltip`.
+   *
+   * This may be removed in a future version of Flarum.
```
Comment (askvortsov1) [inline js/src/@types/global.d.ts:64]:
```suggestion
   * This will be removed in a future version of Flarum.
```

File: js/src/@types/global.d.ts [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+   */
+  tooltip: import('./tooltips/index').TooltipJQueryFunction;
```
Comment (askvortsov1) [inline js/src/@types/global.d.ts:66]: Can we mark this as deprecated?

**Review by askvortsov1 — APPROVED** [review]
Not tested locally, but generally LGTM. Solid set of changes!

Left a few nits.

File: js/src/@types/global.d.ts (hunk 2 above)
Comment (davwheat) [reply js/src/@types/global.d.ts:66]: Most definitely!

File: js/src/@types/global.d.ts (hunk 1 above)
Comment (davwheat) [reply js/src/@types/global.d.ts:64]:
```suggestion
   * This will be removed in a future version of Flarum.
   *
   * @deprecated
```

**Review by SychO9 — APPROVED** [review]

**Comment by SychO9** [comment]
Weird 🤔, compiling with this seems to break extensions that use TS but don't import `app`.
app points to `flarum.core.app`. 

```console
Uncaught TypeError: Cannot read property 'addTranslations' of undefined
Uncaught TypeError: flarum.core.app.load is not a function
```

**Comment by davwheat** [comment]
Hmm... Typing changes couldn't do that. Which extension is this?

**Comment by SychO9** [comment]
never mind, it doesn't have anything to do with extensions, just checking out this branch and building dist files results in the error above. could it be the `../common/app` imports ? how does that point to `Application` ?

**Comment by davwheat** [comment]
It doesn't. In my testing, returning `window.app` seemed to work fine. Might have to revisit that in that case.

**Review by askvortsov1 — APPROVED** [review]

**Review by SychO9 — APPROVED** [review]

**Pull request merged**